ENH: glob for S3 CSVs

TomAugspurger opened the pull request:
@cpcloud this would be super useful for me. Redshift dumps a (gzipped) CSV per node.

My questions:

- priority? I think I took the same as the CSV glob
- Better way to list the keys matching the glob than creating a new S3 connection inside `resource_s3_csv_glob`?
- ~~It doesn't actually work yet... is there another step to hook this up to the network?~~ 
- tests
- is there a better path to hook into than directly from `chunks(s3(CSV)) -> chunks(DataFrame)`?

I've only spent 1-2 hours on this and I'll continue to work on it. I'm just guessing that some of these will be easier for you to answer :)

TomAugspurger commented:
Ah, I see why it wasn't getting added to the graph. `convert` expects (destination, source), not (source, destination). A leftover from `into` I'm guessing.

mrocklin commented:
Indeed

On Fri, Aug 14, 2015 at 12:01 PM, Tom Augspurger <notifications@github.com>
wrote:

> Ah, I see why it wasn't getting added to the graph. convert expects
> (destination, source), not (source, destination). A leftover from into
> I'm guessing.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/ContinuumIO/odo/pull/282#issuecomment-131210244>.
>

cpcloud commented on odo/backends/aws.py:
```diff
@@ -138,6 +138,26 @@ def resource_s3_csv(uri, **kwargs):
 
+@resource.register('s3://.*\*.csv(\.gz)?', priority=19)
+def resource_s3_csv_glob(uri, **kwargs):
+    import boto
```
You can put this at the top level

cpcloud commented:
@TomAugspurger what do you think about generalizing a bit here and making a `Bucket` type?

cpcloud commented:
maybe something like

```python
from odo import Bucket
bucket = resource('s3://path/to/my/bucket/*.csv')  # type(bucket) == Bucket(S3(CSV))
bucket.name == 'bucket'
bucket.path == 'path/to/my/my/bucket'
```

cpcloud commented:
> Better way to list the keys matching the glob than creating a new S3 connection inside `resource_s3_csv_glob`?

Call `get_s3_connection` which memoizes the connection keyed by the requisite AWS credentials

cpcloud commented:
> priority? I think I took the same as the CSV glob

yours should be higher priority, since you want yours to match before matching a plain CSV file

cpcloud commented:
Hm, I wonder if we could make this work with `boto.Bucket` objects...

cpcloud commented:
Ok, scratch the `boto.Bucket` idea for now ... doesn't look like there's a way to make the bucket look as if it has a particular set of files, which we need to do if we wanted to use bucket + globbing.

@TomAugspurger I'm think the chunks approach is nice.

TomAugspurger commented:
Ok, refactored a bit using your comments. I'm reusing the s3 connection, added / tweaked the S3 docs a bit. Tests are still lacking :/ but it works for now at least.


```python
In [1]: import odo

In [2]: odo.odo('s3://vehcuity/tst*.csv.gz', 'test.csv')
Out[2]: <odo.backends.csv.CSV at 0x105a58358>

In [3]: !cat test.csv
0,1
0.07331551328957626,-0.4154916469412773
-1.562069894747674,-1.4722870727590358
-0.15131201641683129,-0.31657923775839325
-0.666743351952952,-1.0564613439645762

In [4]: r = odo.resource('s3://vehcuity/tst*.csv.gz')

In [5]: r
Out[5]: <odo.chunks.chunks(odo.S3(CSV)) at 0x105ae4748>

In [6]: r.
r.container  r.data

In [6]: r.data
Out[6]:
[<odo.backends.aws.S3(CSV) at 0x105ae4390>,
 <odo.backends.aws.S3(CSV) at 0x105ae42b0>]
````

cpcloud commented on odo/backends/aws.py:
```diff
@@ -138,6 +138,37 @@ def resource_s3_csv(uri, **kwargs):
+
+    all_keys = bucket.list(prefix=key.split('*')[0])
+    matched = _filter_to_matched(key, all_keys)
+    uris = [_key_to_uri(k, bucket) for k in matched]
```
i think it's okay to inline `_filter_to_matched` and `_key_to_uri`, IMO it's pretty clear what they are doing

cpcloud commented on odo/backends/aws.py:
```diff
@@ -138,6 +138,37 @@ def resource_s3_csv(uri, **kwargs):
+    bucket = con.get_bucket(result.netloc)
+    key = result.path.lstrip('/')
+
+    all_keys = bucket.list(prefix=key.split('*')[0])
```
awesome! i needed this for https://github.com/blaze/odo/pull/293/files

TomAugspurger commented:
@cpcloud inlined those to functions, I had them broken out for testing, since they don't do any IO, but then gave up on testing :) I need to learn how mock works.

cpcloud commented:
Fixing this up and merging. I've added a test
